File: server/models/Game.js
import mongoose from 'mongoose';

const moveSchema = new mongoose.Schema({
  from: { type: String, required: true },
  to: { type: String, required: true },
  piece: { type: String, required: true },
  captured: { type: String },
  promotion: { type: String },
  san: { type: String, required: true },
  timestamp: { type: Date, default: Date.now }
});

const gameSchema = new mongoose.Schema({
  whitePlayer: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: false  // Agora pode ser null para desafios livres
  },
  blackPlayer: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: false  // Agora pode ser null para desafios livres
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  isOpenChallenge: {
    type: Boolean,
    default: false  // True se for desafio livre (aberto para qualquer um)
  },
  accepted: {
    type: Boolean,
    default: false  // True quando o desafio for aceito
  },
  moves: [moveSchema],
  currentFen: {
    type: String,
    default: 'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1'
  },
  status: {
    type: String,
    enum: ['pendente', 'em_andamento', 'xeque_mate', 'empate', 'abandonada'],
    default: 'pendente'  // Começa como pendente até ser aceito
  },
  winner: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  startedAt: {
    type: Date,
    default: Date.now
  },
  endedAt: {
    type: Date
  },
  result: {
    type: String,
    enum: ['1-0', '0-1', '1/2-1/2', null],
    default: null
  }
});

// Index para buscar jogos de um usuario
gameSchema.index({ whitePlayer: 1, blackPlayer: 1 });
gameSchema.index({ status: 1 });
gameSchema.index({ isOpenChallenge: 1, accepted: 1 });

export default mongoose.model('Game', gameSchema);

File: neumann_chess_flutter/lib/models/game.dart
import 'user.dart';

class Game {
  final String id;
  final User whitePlayer;
  final User blackPlayer;
  final User createdBy;
  final String status;
  final String? result;
  final String? winnerId;
  final String currentFen;
  final String currentTurn;
  final List<Move> moves;
  final bool isOpenChallenge;
  final bool accepted;
  final DateTime? startedAt;
  final DateTime? endedAt;

  Game({
    required this.id,
    required this.whitePlayer,
    required this.blackPlayer,
    required this.createdBy,
    required this.status,
    this.result,
    this.winnerId,
    required this.currentFen,
    required this.currentTurn,
    required this.moves,
    required this.isOpenChallenge,
    required this.accepted,
    this.startedAt,
    this.endedAt,
  });

  factory Game.fromJson(Map<String, dynamic> json) {
    // Helper para converter whitePlayer/blackPlayer/createdBy que podem ser String (ID) ou Map (objeto completo)
    User _parseUser(dynamic userData) {
      if (userData is String) {
        // Se for apenas o ID, criar User com dados mínimos
        return User(
          id: userData,
          username: 'Carregando...',
          email: '',
          stats: UserStats(gamesPlayed: 0, gamesWon: 0, gamesLost: 0, gamesDraw: 0),
        );
      } else if (userData is Map<String, dynamic>) {
        // Se for objeto completo, fazer parse normal
        return User.fromJson(userData);
      } else {
        // Fallback para dados vazios
        return User(
          id: '',
          username: 'Desconhecido',
          email: '',
          stats: UserStats(gamesPlayed: 0, gamesWon: 0, gamesLost: 0, gamesDraw: 0),
        );
      }
    }

    return Game(
      id: json['_id'] ?? json['id'] ?? '',
      whitePlayer: _parseUser(json['whitePlayer']),
      blackPlayer: _parseUser(json['blackPlayer']),
      createdBy: _parseUser(json['createdBy']),
      status: json['status'] ?? 'pendente',
      result: json['result'],
      winnerId: json['winnerId'],
      currentFen: json['currentFen'] ?? 'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1',
      currentTurn: json['currentTurn'] ?? 'white',
      moves: (json['moves'] as List?)?.map((m) => Move.fromJson(m)).toList() ?? [],
      isOpenChallenge: json['isOpenChallenge'] ?? false,
      accepted: json['accepted'] ?? false,
      startedAt: json['startedAt'] != null ? DateTime.parse(json['startedAt']) : null,
      endedAt: json['endedAt'] != null ? DateTime.parse(json['endedAt']) : null,
    );
  }

  // ✅ NOVO: Método copyWith para criar cópias com campos modificados
  Game copyWith({
    String? id,
    User? whitePlayer,
    User? blackPlayer,
    User? createdBy,
    String? status,
    String? result,
    String? winnerId,
    String? currentFen,
    String? currentTurn,
    List<Move>? moves,
    bool? isOpenChallenge,
    bool? accepted,
    DateTime? startedAt,
    DateTime? endedAt,
  }) {
    return Game(
      id: id ?? this.id,
      whitePlayer: whitePlayer ?? this.whitePlayer,
      blackPlayer: blackPlayer ?? this.blackPlayer,
      createdBy: createdBy ?? this.createdBy,
      status: status ?? this.status,
      result: result ?? this.result,
      winnerId: winnerId ?? this.winnerId,
      currentFen: currentFen ?? this.currentFen,
      currentTurn: currentTurn ?? this.currentTurn,
      moves: moves ?? this.moves,
      isOpenChallenge: isOpenChallenge ?? this.isOpenChallenge,
      accepted: accepted ?? this.accepted,
      startedAt: startedAt ?? this.startedAt,
      endedAt: endedAt ?? this.endedAt,
    );
  }
}

class Move {
  final String from;
  final String to;
  final String piece;
  final String? captured;
  final String? promotion;
  final String san;
  final DateTime timestamp;

  Move({
    required this.from,
    required this.to,
    required this.piece,
    this.captured,
    this.promotion,
    required this.san,
    required this.timestamp,
  });

  factory Move.fromJson(Map<String, dynamic> json) {
    return Move(
      from: json['from'] ?? '',
      to: json['to'] ?? '',
      piece: json['piece'] ?? '',
      captured: json['captured'],
      promotion: json['promotion'],
      san: json['san'] ?? '',
      timestamp: json['timestamp'] != null
          ? DateTime.parse(json['timestamp'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'from': from,
      'to': to,
      'piece': piece,
      'captured': captured,
      'promotion': promotion,
      'san': san,
      'timestamp': timestamp.toIso8601String(),
    };
  }
}
